File: 2615. sum-of-distances.py
class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        num_indices = dict()
        occ=dict()
        for i, num in enumerate(nums):
            if num not in num_indices:
                num_indices[num] = i
                occ[num]=1
            else:
                num_indices[num]=num_indices[num]+i
                occ[num]=occ[num]+1
        arr = [0] * len(nums)
        n=len(nums)
        for i in range(n):
            arr[i] = num_indices[nums[i]] - occ[nums[i]]*i     
            num_indices[nums[i]]=num_indices[nums[i]]-2*i
            occ[nums[i]]=occ[nums[i]]-2
            
        return arr

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    nums1 = [1,3,1,1,2]
    output1 = sol.distance(nums1)
    print(f"Example 1: nums={nums1}")
    print(f"Output: {output1}")
    print(f"Expected: [5,0,3,4,0]")
    print()
    
    # Example 2
    nums2 = [0,5,3]
    output2 = sol.distance(nums2)
    print(f"Example 2: nums={nums2}")
    print(f"Output: {output2}")
    print(f"Expected: [0,0,0]")
        

File: 2615. sum-of-distances.java
import java.util.ArrayList;
import java.util.HashMap;

class Solution {
    public long[] distance(int[] nums) {
        HashMap<Integer, ArrayList<Integer>> hm = new HashMap<>();
        for(int i = 0; i < nums.length; i++) {
            if(!hm.containsKey(nums[i])) {
                hm.put(nums[i], new ArrayList<>());
            }
            hm.get(nums[i]).add(i);
        }
        
        long[] ans = new long[nums.length];
        for(ArrayList<Integer> indices : hm.values()) {
            int n = indices.size();
            long sum = 0;
            for(int i = 0; i < n; i++) {
                sum += indices.get(i);
            }
            long leftsum = 0;
            long rightsum = sum;
            long currsum = 0;
            for(int i = 0; i < n; i++) {
                currsum = 0;
                currsum += (long)i*(long)indices.get(i) - leftsum;
                currsum += rightsum - (long)(n-i)*(long)indices.get(i);
                leftsum += (long)indices.get(i);
                rightsum -= (long)indices.get(i);
                ans[indices.get(i)] = currsum;
            }
        }
        return ans;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[] nums1 = {1,3,1,1,2};
        long[] output1 = sol.distance(nums1);
        System.out.println("Example 1: nums=[1,3,1,1,2]");
        System.out.print("Output: [");
        for(int i = 0; i < output1.length; i++) {
            if(i > 0) System.out.print(",");
            System.out.print(output1[i]);
        }
        System.out.println("]");
        System.out.println("Expected: [5,0,3,4,0]");
        System.out.println();
        
        // Example 2
        int[] nums2 = {0,5,3};
        long[] output2 = sol.distance(nums2);
        System.out.println("Example 2: nums=[0,5,3]");
        System.out.print("Output: [");
        for(int i = 0; i < output2.length; i++) {
            if(i > 0) System.out.print(",");
            System.out.print(output2[i]);
        }
        System.out.println("]");
        System.out.println("Expected: [0,0,0]");
    }
}